web/datapipeline/sap_systems_projector.go

```diff
@@ -0,0 +1,81 @@
+package datapipeline
+
+import (
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/trento-project/trento/internal/sapsystem"
+	"github.com/trento-project/trento/web/entities"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+const SAPSystemsApplication = "application"
+const SAPSystemsDatabase = "database"
+
+func NewSAPSystemsProjector(db *gorm.DB) *projector {
+	SAPSystemsProjector := NewProjector("sapsystems", db)
+
+	SAPSystemsProjector.AddHandler(SAPsystemDiscovery, SAPSystemsProjector_SAPSystemsDiscoveryHandler)
+
+	return SAPSystemsProjector
+}
+
+func SAPSystemsProjector_SAPSystemsDiscoveryHandler(dataCollectedEvent *DataCollectedEvent, db *gorm.DB) error {
+	decoder := getPayloadDecoder(dataCollectedEvent.Payload)
+	var discoveredSAPSystems sapsystem.SAPSystemsList
+	if err := decoder.Decode(&discoveredSAPSystems); err != nil {
+		log.Errorf("can't decode data: %s", err)
+		return err
+	}
+
+	for _, s := range discoveredSAPSystems {
```

Comment:
I see some issues in this loop. 
Here, we are looping the SAP systems installed in 1 host. There might be 1..n system.
But within the SAP system, we can have multiple SAP instances installed, 1..n again. So, I miss an additional loop here, which creates the SAP instance entry itself.

Example: We have a netweaver installation that has the ASCS and PAS in the same machine. Both belong to the same SAP system, but are 2 different instance (2 rows in the UI table).

So most probably, we would need to run the `storeSAPInstance` function within the `s.Instaces` loop.

That's why I see that the instances and systems are kind of coupled. But I guess we can introduce the system data in the instance as well (as they are parent/children somehow)

Reply:
> That's why I see that the instances and systems are kind of coupled

Absolutely, I see the point but I wanted to do this at service level, populating a `SAPSystem` model grouping many `SAPSystemsInstance` entities cast into proper models.

> Example: We have a netweaver installation that has the ASCS and PAS in the same machine. Both belong to the same SAP system, but are 2 different instance (2 rows in the UI table).

Understood, I'll try to come up with something better, thanks for the feedback